src/Cli/dotnet/commands/Run/VirtualProjectBuildingCommand.cs

```diff
@@ -139,86 +178,443 @@ private ProjectInstance CreateProjectInstance(
         {
             GlobalProperties = globalProperties,
         });
+
+        ProjectRootElement CreateProjectRootElement(ProjectCollection projectCollection)
+        {
+            Debug.Assert(!_directives.IsDefault && _targetFilePath is not null, $"{nameof(PrepareProjectInstance)} should have been called first.");
+
+            var projectFileFullPath = Path.ChangeExtension(EntryPointFileFullPath, ".csproj");
+            var projectFileWriter = new StringWriter();
+            WriteProjectFile(projectFileWriter, _directives, virtualProjectFile: true, targetFilePath: _targetFilePath);
+            var projectFileText = projectFileWriter.ToString();
+
+            using var reader = new StringReader(projectFileText);
+            using var xmlReader = XmlReader.Create(reader);
+            var projectRoot = ProjectRootElement.Create(xmlReader, projectCollection);
+            projectRoot.FullPath = projectFileFullPath;
+            return projectRoot;
+        }
     }
 
-    // Kept in sync with the default `dotnet new console` project file (enforced by `DotnetProjectAddTests.SameAsTemplate`).
-    private const string CommonProjectProperties = """
-            <OutputType>Exe</OutputType>
-            <TargetFramework>net10.0</TargetFramework>
-            <ImplicitUsings>enable</ImplicitUsings>
-            <Nullable>enable</Nullable>
-        """;
+    public static void WriteProjectFile(TextWriter writer, ImmutableArray<CSharpDirective> directives)
+    {
+        WriteProjectFile(writer, directives, virtualProjectFile: false, targetFilePath: null);
+    }
 
-    public static string GetNonVirtualProjectFileText()
+    private static void WriteProjectFile(TextWriter writer, ImmutableArray<CSharpDirective> directives, bool virtualProjectFile, string? targetFilePath)
```

Comment:
nit: consider a more conventional name for a flag like `isVirtualProject`.

Also, I am wondering if the differences when project file is virtual, are so many/complex, that we would be better off with copy/pasted versions of this function, which each have a more clear flow of logic. Up to you whether such a change would be worthwhile.

Reply:
> Also, I am wondering if the differences when project file is virtual, are so many/complex, that we would be better off with copy/pasted versions of this function, which each have a more clear flow of logic. Up to you whether such a change would be worthwhile.

I prefer the current approach. We would need to have lot of duplicated logic and the virtual project differences are mostly self-contained `if`s apart from the sdk directives, so hopefully it isn't too confusing. Also ideally the workarounds will be removed some time in the future and both virtual and physical projects will look the same.